docs: enhance assert_changes and assert_no_changes documentation [ci skip]

saiqulhaq opened the pull request:
:link: [Preview](https://saiqulhaq-document-case-equa.rails-docs-preview.pages.dev/api/classes/ActiveSupport/Testing/Assertions.html#method-i-assert_changes)

### Detail

This Pull Request has been created to fix issue number #53693 
Added comprehensive examples demonstrating pattern-matching and type-checking capabilities of the assert_changes and assert_no_changes methods.

MatheusRich commented on activesupport/lib/active_support/testing/assertions.rb:
```diff
@@ -181,12 +181,24 @@ def assert_no_difference(expression, message = nil, &block)
       #     @object = :foo
       #   end
       #
+      #   # Pattern match
```
I think it's better to change the wording here to avoid confusion with [Ruby pattern matching](https://docs.ruby-lang.org/en/3.4/syntax/pattern_matching_rdoc.html)
```suggestion
      #   # Case equality
```

saiqulhaq replied:
Thank you for the feedback! You're absolutely right - we should avoid conflating the `===` semantics with Ruby's formal pattern-matching syntax.

saiqulhaq committed:
docs: enhance assert_changes and assert_no_changes documentation

Added comprehensive examples demonstrating case equality and type checking
capabilities of the assert_changes and assert_no_changes methods.
The examples show how these methods can be used with exact value matches,
regular expressions, and class type checks using the case equality (===) operator.
